drivers/tty/serdev/pibridge.c

```diff
@@ -0,0 +1,409 @@
+{
+	struct pibridge *pi = serdev_device_get_drvdata(serdev);
+
+	serdev_device_close(serdev);
+
+	kfifo_free(&pi->read_fifo);
+};
+
+#ifdef CONFIG_OF
+static const struct of_device_id pibridge_of_match[] = {
+	{ .compatible = "kunbus,pibridge" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, pibridge_of_match);
+#endif
+
+static struct serdev_device_driver pibridge_driver = {
+	.driver	= {
+		.name		= "pibridge",
+		.of_match_table	= of_match_ptr(pibridge_of_match),
+	},
+	.probe	= pibridge_probe,
+	.remove	= pibridge_remove,
+};
+module_serdev_device_driver(pibridge_driver);
+
+
+/*****************/
+
+int pibridge_send(u8 *buf, u16 len)
+{
+	struct pibridge *pi = pibridge_s;
+	struct serdev_device *serdev = pi->serdev;
+	int ret;
+
+	ret = serdev_device_write(serdev, buf, len, MAX_SCHEDULE_TIMEOUT);
+	if (ret < 0)
+		pb_err(25, "write to serdev error(len:%d)\n", len);
+
+	/* read fifo may contain stale bytes */
+	mutex_lock(&pi->lock);
+	kfifo_reset(&pi->read_fifo);
+	mutex_unlock(&pi->lock);
+
+	serdev_device_wait_until_sent(serdev, 0);
+
+	return ret;
+}
+
+int pibridge_recv_timeout(u8 *buf, u16 len, u16 timeout)
+{
+	struct pibridge *pi = pibridge_s;
+	int jiffies;
+	int ret;
+
+	jiffies = wait_event_timeout(pi->read_queue, kfifo_len(&pi->read_fifo) >= len,
+			msecs_to_jiffies(timeout));
+	mutex_lock(&pi->lock);
+	if (jiffies){
+		ret = kfifo_out(&pi->read_fifo, buf, len);
+	} else {
+		ret = 0;
+	}
+	mutex_unlock(&pi->lock);
+	if (ret < len)
+		pb_err(25, "receive message error(len:%d, ret:%d, jiffies:%d, fifo:%d)\n",
+				len, ret, jiffies, kfifo_len(&pi->read_fifo));
+	return ret;
+}
+
+int pibridge_recv(u8 *buf, u16 len)
+{
+	/* using default timeout REV_PI_IO_TIMEOUT */
+	return pibridge_recv_timeout(buf, len, REV_PI_IO_TIMEOUT);
+}
+
+#pragma pack(1)
+struct pibridge_pkthdr_gate{
+	u8	dst;
+	u8	src;
+	u16	cmd;
+	u16	seq;
+	u8	len;
+};
+
+u8 pibridge_crc8(u8 base, u8 *data, u16 len)
+{
+	u8 ret = base;
+
+	while (len--) {
+		ret = ret ^ data[len];
+	}
+	return ret;
+}
+
+int pibridge_req_send_gate(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len)
+{
+	struct pibridge_pkthdr_gate pkthdr;
+	u8 crc;
+
+	pkthdr.dst = dst;
+	pkthdr.src = 0;
+	pkthdr.cmd = cmd;
+	pkthdr.seq = 0;
+	pkthdr.len = snd_len;
+
+	if (pibridge_send((u8 *)&pkthdr, sizeof(pkthdr))){
+		pb_err(25, "send head error in gate-send\n");
+		return -EIO;
+	}
+	if (snd_len != 0){
+		if (pibridge_send(snd_buf, snd_len)){
+			pb_err(25, "send data error in gate-send(len:%d)\n",
+					snd_len);
+			return -EIO;
+		}
+
+	}
+
+	crc = pibridge_crc8(0, (u8 *)&pkthdr, sizeof(pkthdr));
+	if (snd_len != 0){
+		crc = pibridge_crc8(crc, snd_buf, snd_len);
+	}
+
+	if (pibridge_send(&crc, sizeof(u8))){
+		pb_err(25, "send crc error in gate-send\n");
+		return -EIO;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL(pibridge_req_send_gate);
+
+#define PIBRIDGE_RESP_CMD	0x3fff
+#define PIBRIDGE_RESP_OK	0x4000
+#define PIBRIDGE_RESP_ERR	0x8000
+
+int pibridge_req_gate_tmt(u8 dst, u16 cmd, u8 *snd_buf, u16 snd_len,
```

Comment:
For now, passing a destination address to these functions and having a global `pibridge_s` instance is fine.

Once we perform a bus scan and instantiate a `struct pibridge_dev` for each module found, we'll pass that struct to the send/receive functions and it'll contain both the module address and the pointer to the `struct pibridge`.